tools/cdb2_sqlreplay/cdb2_sqlreplay.cpp

```diff
@@ -525,38 +530,47 @@ void replay(cdb2_hndl_tp *db, cson_value *event_val) {
      eliminate it.
   */
 
-void handle_sql(cdb2_hndl_tp *db, cson_value *event_val) {
-    int rc;
-    /* We can only replay if we have the full SQL, including parameters.
-       That means
-       1) event logged a bindings array
-       2) event logged a nbindings integer, and it's 0
-
-       If there's non-zero bindings and we don't have them, we can't replay.
 
-       Another complication is transactions - we don't want to play one back
-       until we have all the statements collected, and see a commit event.
-       Anything with an id logged is part of a transaction - a commit
-       event will have the same id.  SQL writes that are not part of
-       a transaction get logged after transaction.  SQL in BEGIN/COMMIT
-       blocks gets logged before the transaction.
+/* We can only replay if we have the full SQL, including parameters.
+   That means
+   1) event logged a bindings array
+   2) event logged a nbindings integer, and it's 0
 
-       */ 
+   If there's non-zero bindings and we don't have them, we can't replay.
 
+   Another complication is transactions - we don't want to play one back
+   until we have all the statements collected, and see a commit event.
+   Anything with a cnonce logged is part of a transaction - a commit
+   event will have the same cnonce.  SQL writes that are not part of
+   a transaction get logged before txn log entry.  SQL in BEGIN/COMMIT
+   blocks gets logged before the txn log entry as well.
+   */ 
+void handle_sql(cdb2_hndl_tp *db, cson_value *event_val) {
+    int rc;
     if (!is_replayable(event_val)) {
         cson_free_value(event_val);
         return;
     }
 
-#if 0
-    if (is_transactional(event_val)) {
-        add_to_transaction(db, event_val);
+    const char *cnonce = get_strprop(event_val, "cnonce");
+    if (cnonce == nullptr) { // no cnonce if comdb2api -- just replay it
+        replay(db, event_val);
+        cson_free_value(event_val);
         return;
     }
-#endif
-    
-    replay(db, event_val);
-    cson_free_value(event_val);
+
+    const char *sql = get_strprop(event_val, "sql");
+    if (strcmp(sql, "commit") == 0 && is_part_of_transaction(cnonce)) {
+        add_to_transaction(event_val);
+    } else if (strcmp(sql, "rollback") == 0) {
+        replay_transaction(db, event_val);
```

Comment:
Why do we replay on rollback?

Reply:
We do so to get cost of reads within the transaction, otherwise we would lack cost info for those querie on replays.
The idea is to compare one version of the db with another and find regressions on all possible queries.

Reply:
I suppose we can have cmdline parameters to omit those, but for now I think this is what we need.The replayed eventlog can be compared directly and should be almost identical (in order of queries), for instance for the replay_events.test/ 
```
> wc -l events.unzipped secondary_events.unzipped 
   10301 events.unzipped
   10299 secondary_events.unzipped
```